### PR TITLE
PE-1412: track yet not unbundled bundles to aggregate on the next run

### PR DESCRIPTION
**Pull request by matibat**

### CODE REVIEW

**Comment by matibat** [comment]
Todo: setup the bundle tips before aggregating the bundled data

File: .github/workflow/aggregate.yaml [hunk 1]
```diff
@@ -40,3 +40,13 @@ jobs:
 
       - name: Publish output
         run: ardrive upload-file --local-path ./daily_output.json -F "d62a47c3-0b9d-4442-ac72-252a239d0469" -w /tmp/wallet --no-bundle --upsert --dry-run
+
+      - name: Read non unbundled bundles
```
Comment (karlprieb) [inline .github/workflow/aggregate.yaml]: Where are we using this env variable?

File: src/queries.ts [hunk 2]
```diff
@@ -80,6 +80,34 @@ export async function getAllArDriveTransactionsWithin(minBlock: number, maxBlock
...
+	// const blockHeight = await getBlockHeight();
+
+	let hasNextPage = true;
+
+	while (hasNextPage) {
+		const query = createBundledTxsQuery(txIds);
+		const response = await sendQuery(query);
+		if (response.edges.length) {
+			console.log(`Transactions count: ${response.edges.length}`);
+			// const mostRecentTransaction = response.edges[response.edges.length - 1];
```
Comment (karlprieb) [inline src/queries.ts]: can we remove theses comments?

File: src/queries.ts [hunk 3]
```diff
@@ -190,3 +218,56 @@ function createQuery(minBlock: number, maxBlock: number): Query {
 			}`
 	};
 }
+
+/**
+ */
+function createBundledTxsQuery(txIDs: string[]): Query {
+	// FIXME: use minBlock & maxBlock
```
Comment (karlprieb) [inline src/queries.ts:210]: 👍

File: .github/workflow/aggregate.yaml (hunk 1 above)
Comment (matibat) [reply .github/workflow/aggregate.yaml]: Nowhere. e4701eb127983b86021e035fa41f12e4b617e907

File: src/queries.ts (hunk 2 above)
Comment (matibat) [reply src/queries.ts]: e8934ce85da42872a47d21bbec3bd1c3856a7d34

File: src/daily_output.ts [hunk 4]
```diff
@@ -66,13 +68,37 @@ export class DailyOutput {
...
 	 * - streak rewards
 	 */
-	private finishDataAggregation(): void {
+	private async finishDataAggregation(): Promise<void> {
+		// check for previous unbundled bundles
+		const prevNonUnbundledBundles: { txId: string; tip: number }[] = existsSync('non_unbundled_bundles.json')
+			? JSON.parse(readFileSync('non_unbundled_bundles.json').toString())
+			: [];
+		const unbundledTransactions = prevNonUnbundledBundles.length
+			? await getBundledTransactions(prevNonUnbundledBundles.map(({ txId }) => txId))
```
Comment (fedellen) [inline src/daily_output.ts]: I think you'll want `feedGQLData` to become a promise once again here to do this. Sorry for having you refactor that, seems it was meant to be a promise

Comment (fedellen) [reply src/daily_output.ts]: You could also consider a class that handles tracking unbundled bundles and uses `fs`

File: src/daily_output.ts [hunk 5]
```diff
@@ -66,13 +68,37 @@ export class DailyOutput {
 	 * - group effort rewards, and
 	 * - streak rewards
 	 */
-	private finishDataAggregation(): void {
+	private async finishDataAggregation(): Promise<void> {
+		// check for previous unbundled bundles
+		const prevNonUnbundledBundles: { txId: string; tip: number }[] = existsSync('non_unbundled_bundles.json')
```
Comment (fedellen) [inline src/daily_output.ts]: Let's extract that file name to a constant

File: src/daily_output.ts [hunk 6]
```diff
@@ -66,13 +68,37 @@ export class DailyOutput {
...
+		const unbundledTransactions = prevNonUnbundledBundles.length
+			? await getBundledTransactions(prevNonUnbundledBundles.map(({ txId }) => txId))
+			: [];
+		// track the tips of the bundles yet not unbundled at the previous run
+		prevNonUnbundledBundles.forEach(({ tip, txId }) => {
+			this.bundlesTips[txId] = tip;
+		});
+		unbundledTransactions.forEach(this.aggregateData);
+
+		// calculate this run's non unbubdled bundles
```
Comment (fedellen) [inline src/daily_output.ts]:
```suggestion
		// calculate this run's non unbundled bundles
```

**Review by fedellen — APPROVED** [review]
I think we're good to go here

File: src/daily_output.ts (hunk 4 above)
Comment (matibat) [reply src/daily_output.ts]: 79807f400c121c79fbbf742f737d4413bd09162d

File: src/daily_output.ts (hunk 5 above)
Comment (matibat) [reply src/daily_output.ts]: cbcfb0e571c7e90691ccb81d772ade4edaf41451

**Review by fedellen — APPROVED** [review]

**Pull request merged**